R/loop_tasks.R

```diff
@@ -114,57 +120,96 @@ loop_tasks <- function(
         this_try, num_tries, num_targets_incomplete, num_targets_overall)
       pb$message(loop_start_msg)
     }
-    this_try <- this_try + 1
-    
-    # prepare a vector to record successes/failures within this loop
-    target_succeeded <- rep(FALSE, num_targets_incomplete)
     
-    # attempt to build each of the incomplete targets
-    for(i in seq_len(num_targets_incomplete)) {
-      tryCatch({
-        # get the names of the target and the task
-        target_num_overall <- incomplete_targets[i]
-        target <- targets[target_num_overall]
-        task_name <- task_names[target_num_overall]
-        
-        # update the progress bar
-        if (verbose){
-          pb$update(
-            num_targets_complete/num_targets_overall, 
-            tokens = list(what = sprintf('  Building %s', task_name)))
-        }
-        
-        # the main action: run the task-step
-        scmake(target, task_makefile, ind_ext=ind_ext, verbose=FALSE)
-        
-        # if it worked, note the success
-        target_succeeded[i] <- TRUE
-        num_targets_complete <- num_targets_complete + 1
-        
-      }, error=function(e) {
-        if(verbose){
-          pb$message(sprintf(
-            "* Error in %s: %s; debug with scmake(\"%s\", \"%s\")", 
-            deparse(e$call), e$message, target, task_makefile))
-        }
-        
-        # sleep for a while if requested
-        if(sleep_on_error > 0) {
-          Sys.sleep(sleep_on_error)
-        }
-      })
-
-      # try to keep memory under control if possible; might be harder with
-      # object targets, not sure if storr keeps them all in memory
-      gc()
+    error_function <- function(e) {
+      if(verbose){
+        pb$message(sprintf(
+          "* Error in %s: %s; debug with scmake(\"%s\", \"%s\")", 
+          deparse(e$call), e$message, target, task_makefile))
+      }# sleep for a while if requested
+      if(sleep_on_error > 0) {
+        Sys.sleep(sleep_on_error)
+      }
     }
-    
-    # revise and recount the list of incomplete targets for the next while loop iteration
-    incomplete_targets <- incomplete_targets[!target_succeeded]
-    num_targets_incomplete <- length(incomplete_targets) # count of remaining targets to try in this loop
+    #would parallelize here
+    if(!parallel) {
+      # prepare a vector to record successes/failures within this loop
+      target_succeeded <- rep(FALSE, num_targets_incomplete)
+      # attempt to build each of the incomplete targets
+      for(i in seq_len(num_targets_incomplete)) {
+        tryCatch({
+          # get the names of the target and the task
+          target_num_overall <- incomplete_targets[i]
+          target <- targets[target_num_overall]
+          task_name <- task_names[target_num_overall]
+          
+          # update the progress bar
+          #probably can't do this parallel?  Would result in race conditions?
```

Comment:
Yeah, as is right now the progress bar is just updated after the `foreach` loop is done, so unless there are task failures it will just go straight to 100%

Reply:
Sounds good to me.